accepted/00000-maintenance-windows.md

```diff
@@ -0,0 +1,191 @@
+to a different time. Possible modules could be:
+
+- Nearest: Use the nearest maintenance window
+- Previous: Use the maintenance window before the scheduled date
+- Next: Use the maintenance window after the scheduled date
+- Cancel: Cancel the Action
+- Fail: do not import the new schedule at all and keep the existing one
+
+The customer can configure the order:
+
+  resolveOrder=Nearest,Previous,Next,Fail
+
+
+### Schedule an Action
+
+When a system is assigned to a schedule, only schedules withing a maintenance window should be allowed.
+In the User Interface we should show in these cases not the datepicker, but two simple Combo boxes with Date and Time.
+The Date contains the next 20 dates where a maintenance window exists. Depending on the date box the available
+times in the time box change.
+
+When using SSM it is possible that systems have different maintenance windows. The combination systems belong
+to one maintenance schedule and systems without maintenance schedule is not critical. The action would be scheduled
+at a time which is inside of a maintenance window of the systems which belong to the schedule.
+
+Systems in different schedules needs special handling:
+
+1. do not allow systems with different schedules in SSM
+2. divide systems by maintenance schedules and display a combo box with the names. Let the user choose for which set he want to schedule it. He need to repeat it for the rest.
+3. show the picker multiple times; for every maintenance schedule one.
+
+(1 limiting too much and 3 might look ugly - I would prefer Option 2)
+
+
+### Assign a Maintenance Schedule to a System
+
+There could be multiple ways to assign systems to maintenance schedules.
+
+- In the systems details page we need to show the selected schedule with a view and a change option.
+- It should be possible to change the maintenance schedule for multiple systems using SSM
+- In the details of the Maintenance Schedule we could have a systems page where add/remove is possible.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+
+# Alternatives
+[alternatives]: #alternatives
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+
+## What does it mean to turn maintenance mode "on" and "off"
+
+- A flag in DB which needs to be changed?
+- We evaluate the calendar and calculate the state of the system?
+- mix: taskomatic job (every minute) which evaluate the calendar and switch the flag (and trigger any other action which is needed)
+
+Maybe the mix, but maintenance mode it moved to 2nd iteration.
+
+
+## Traditional Clients
+
+They may ask for new jobs only every 2 - 6 hours. Then the maintenance window might be over.
+What to do?
+
+- Execute even after MU is over? (this is what may happen without changes)
+- New features are made for salt - just mention it in the docs and suggest to use osad, a short checking frequency or a very long Window?
+- Special implementation?
+
```

Comment:
My guidance for traditional clients is: Focus on Salt, but if there's no practical difference, I'm super-fine with getting the traditional stack support for free.

_If_ we end up being able traditional clients, my expectation (to be validated by someone in the field) would be that Uyuni strictly enforces the windows. So if a client checks in too late, bad luck and come back in the next window.

That said we'd most likely be able to just point to OSAD.

BTW, one thing we may want to check is whether admins are _currently_ using the run_check to implement their own "maintenance windows" (by only running the run_check when they actually want to do maintenance).

Reply:
This is similar to what I think. When customer uses osad, it should just work. We schedule for a time in the maintenance window and osad will trigger at that time.

Only when not using osad possible problems might happen.
But with documentation and long enough maintenance windows it should be doable.